lib/retry.c

```diff
@@ -176,6 +152,29 @@ EXPORTED ssize_t retry_writev(int fd, const struct iovec *srciov, int iovcnt)
         written += n;
 
         if (written == len) break;
+
+        /* Oh well, welcome to the slow path - we have copies */
+        if (!baseiov) {
+            iov = baseiov = xmalloc(iovcnt * sizeof(struct iovec));
+            for (i = 0; i < iovcnt; i++) {
+                iov[i].iov_base = srciov[i].iov_base;
+                iov[i].iov_len = srciov[i].iov_len;
+            }
+        }
+
+        /* Skip any iov that may have been written in full */
+        while ((size_t) n >= iov->iov_len) {
+            n -= iov->iov_len;
+            iov++;
+            iovcnt--;
+            if (!iovcnt) fatal("ran out of iov", EX_SOFTWARE);
+        }
+
+        /* Skip whatever portion of the current iov that has been written */
+        iov->iov_base += n;
+        iov->iov_len -= n;
+
+        srciov = iov;
```

Comment:
I feel like we have too many different versions of iov here...

Reply:
How so?  iov is our malloc'd copy which we can edit.  srciov always points to what we are trying to output.
I have another version which keeps the initial write separate from the rest, and simply resets n to zero on failure